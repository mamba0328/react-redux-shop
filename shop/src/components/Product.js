import Button from './Buttons'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faStar } from '@fortawesome/free-solid-svg-icons'
import PropTypes from 'prop-types';
import { useNavigate } from 'react-router-dom';

const Product = (props) => {
    const additionalClass = props.additionalClass ? props.additionalClass : "";
    const deckImg = props.deckImg;
    const { name, price, img, imgShort, code, color } = props.product;
    const navigate = useNavigate();

    const handleClick = (code) => {
        navigate('/product/' + code, { replace: true });
    }
    return (
        <li className={'product' + ' ' + additionalClass} onClick={() => handleClick(code)}>
            <div className='product__img-container'>
                <img src={deckImg ? imgShort : img} className='product__img' alt='product img' />
                <img src={deckImg ? imgShort : img} className='product__img product__img_shadow shadow' alt='product img' />

            </div>
            <div className='product__info'>
                {/* <div className='flex-row'>
                    <p className='product__code'>{code}</p>
                    <button className='product__favorite' onClick={(e) => props.toggleFavorite(code, e)}>
                        <FontAwesomeIcon icon={faStar} className={props.isFavorite ? 'product__star checked' : 'product__star'} />
                    </button>
                </div> */}
                <h4 className='product__name'>{name}</h4>
                {/* <p className='product__color'>{color}</p> */}
                <div className='product__rating rating'>
                    <FontAwesomeIcon icon={faStar} />
                    <FontAwesomeIcon icon={faStar} />
                    <FontAwesomeIcon icon={faStar} />
                    <FontAwesomeIcon icon={faStar} />
                    <FontAwesomeIcon icon={faStar} />
                </div>
                <h4 className='product__price'>{'â‚´' + price.split(' ').join(',')}</h4>

                {/* {props.xButton ? <button className='xButton' data-testid='closeBtn' onClick={(e) => props.removeFromCart(code, e)}></button> : props.inCart ? <Button text='In cart' backgroundColor='#e74c3c' /> : <Button text='Add to cart' backgroundColor='#262626' onClick={(e) => props.addToCart(code, e)} />} */}
            </div>
        </li>
    )
}


Product.propTypes = {
    product: PropTypes.object.isRequired,
    toggleFavorite: PropTypes.func.isRequired,
    isFavorite: PropTypes.bool,
    addToCart: PropTypes.func.isRequired,
    removeFromCart: PropTypes.func,
    inCart: PropTypes.bool,
    xButton: PropTypes.bool,
}

Product.defaultProps = {
    isFavorite: false,
    inCart: false,
    xButton: false,
    removeFromCart: () => console.log('can`t remove')
}

export default Product