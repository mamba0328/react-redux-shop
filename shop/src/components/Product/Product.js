import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faStar } from '@fortawesome/free-solid-svg-icons'
import PropTypes from 'prop-types';
import { useNavigate } from 'react-router-dom';

const Product = (props) => {
    const additionalClass = props.additionalClass ? props.additionalClass : "";
    const deckImg = props.deckImg;
    const imageIsLoaded = props.imageIsLoaded ?? function () { }
    const { name, price, code, } = props.product;
    const navigate = useNavigate();

    const fullImageURL = `./png/${code}_full.png`;
    const deckImgURL = `./png/${code}_deck.png`;
    const imgURL = deckImg ? deckImgURL : fullImageURL;

    const handleClick = (code) => {
        navigate('/product/' + code, { replace: true });
    }
    return (
        <li className={'product' + ' ' + additionalClass} onClick={() => handleClick(code)}>
            <div className='product__img-container'>
                <img src={imgURL} onLoad={() => imageIsLoaded(imgURL)} className='product__img product__img_shadow shadow' alt='product img' />
                <img src={imgURL} onLoad={() => imageIsLoaded(imgURL + '2')} className='product__img' alt='product img' />

            </div>
            <div className='product__info'>
                <h4 className='product__name'>{name}</h4>
                <div className='product__rating rating'>
                    <FontAwesomeIcon icon={faStar} />
                    <FontAwesomeIcon icon={faStar} />
                    <FontAwesomeIcon icon={faStar} />
                    <FontAwesomeIcon icon={faStar} />
                    <FontAwesomeIcon icon={faStar} />
                </div>
                <h4 className='product__price'>{'â‚´' + price.split(' ').join(',')}</h4>
            </div>
        </li>
    )
}


Product.propTypes = {
    product: PropTypes.object.isRequired,
    toggleFavorite: PropTypes.func.isRequired,
    isFavorite: PropTypes.bool,
    addToCart: PropTypes.func.isRequired,
    removeFromCart: PropTypes.func,
    inCart: PropTypes.bool,
    xButton: PropTypes.bool,
}

Product.defaultProps = {
    isFavorite: false,
    inCart: false,
    xButton: false,
    removeFromCart: () => console.log('can`t remove')
}

export default Product