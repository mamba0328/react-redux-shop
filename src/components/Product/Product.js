import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faStar, faCartPlus, faCheck } from '@fortawesome/free-solid-svg-icons'
import PropTypes from 'prop-types';
import { useNavigate } from 'react-router-dom';
import { useSelector } from 'react-redux';

const Product = (props) => {
    const additionalClass = props.additionalClass ? props.additionalClass : "";
    const deckImg = props.deckImg;
    const imageIsLoaded = props.imageIsLoaded ?? function () { }
    const { name, price, code, } = props.product;
    const { addToCart, removeFromCart } = props;

    const cart = useSelector(state => state.rootReducer.cart);
    const inCart = cart.includes(code)
    const navigate = useNavigate();

    const fullImageURL = `./png/${code}_full.png`;
    const deckImgURL = `./png/${code}_deck.png`;
    const imgURL = deckImg ? deckImgURL : fullImageURL;

    const handleClick = (code) => {
        navigate('/product/' + code, { replace: true });
    }

    return (
        <li className={'product' + ' ' + additionalClass} onClick={() => handleClick(code)}>
            <div className='product__img-container'>
                <img src={imgURL} onLoad={() => imageIsLoaded(imgURL)} className='product__img product__img_shadow shadow' alt='product img' />
                <img src={imgURL} onLoad={() => imageIsLoaded(imgURL + '2')} className='product__img' alt='product img' />
            </div>
            <div className='product__info'>
                <h4 className='product__name'>{name}</h4>
                <div className='product__rating rating'>
                    <FontAwesomeIcon icon={faStar} />
                    <FontAwesomeIcon icon={faStar} />
                    <FontAwesomeIcon icon={faStar} />
                    <FontAwesomeIcon icon={faStar} />
                    <FontAwesomeIcon icon={faStar} />
                </div>
                <h4 className='product__price'>{'â‚´' + price.split(' ').join(',')}</h4>
                {inCart ? <FontAwesomeIcon icon={faCheck} className='product__icon' onClick={(e) => removeFromCart(code, e)} /> : <FontAwesomeIcon icon={faCartPlus} className='product__icon product__icon_add-to-cart' onClick={(e) => addToCart(code, e)} />}
            </div>
        </li>
    )
}


Product.propTypes = {
    product: PropTypes.object.isRequired,
    toggleFavorite: PropTypes.func.isRequired,
    isFavorite: PropTypes.bool,
    addToCart: PropTypes.func.isRequired,
    removeFromCart: PropTypes.func,
    inCart: PropTypes.bool,
    xButton: PropTypes.bool,
}

Product.defaultProps = {
    isFavorite: false,
    inCart: false,
    xButton: false,
    removeFromCart: () => console.log('can`t remove')
}

export default Product