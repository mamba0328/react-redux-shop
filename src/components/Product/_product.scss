.product {
    display: flex;
    flex-direction: column;
    position: relative;
    min-width: 200px;
    max-width: 200px;
    border-radius: 25px;
    background-color: $card;
    cursor: pointer;
    direction: ltr;

    &:hover {
        .product__img-container {
            transform: translateY(-15%) rotate(-5deg);
        }

        .product__icon {
            display: inline-block;
        }
    }


    &__icon {
        position: absolute;
        right: 15px;
        bottom: 15px;
        display: none;
        transition: all 0.3s ease-in-out;

        &_add-to-cart {
            &:hover {
                transform: scale(1.5);
            }
        }
    }


    &__img-container {
        min-height: 200px;
        width: 100%;
        position: relative;
        transition: all 0.3s ease-in-out;

        -webkit-user-select: none;
        -khtml-user-select: none;
        -moz-user-select: none;
        -o-user-select: none;
        user-select: none;

    }

    &__img {
        position: absolute;
        object-fit: cover;
        width: 100%;
        z-index: 1;
    }

    &__info {
        @include flex(column, center, flex-start);
        gap: 10px;
        padding: 15px;
        padding-top: 0;
    }

    &__img-container,
    &__info {
        transform: translateY(-15%);
    }


    &__name,
    &__price {
        font-weight: 500;
    }

    &__name {
        overflow: hidden;
        display: -webkit-box;
        -webkit-line-clamp: 2;
        line-clamp: 2;
        -webkit-box-orient: vertical;

    }


    &_horizontal {
        flex-direction: row;
        background-color: transparent;

        &:hover {
            .product__img-container {
                transform: translateY(-15%);
            }

            .product__icon {
                display: none;
            }
        }

        & .product__img-container {
            min-width: 80px;
            width: 80px;
            min-height: auto;
            height: 100px;
            border-radius: 10px;
            background-color: $card;
        }

        & .product__img {
            width: 60px;
            left: 50%;
            transform: translateX(-50%);
        }

        & .product__info {
            justify-content: space-between;
            padding-top: 5px;
            padding-bottom: 5px;
        }
    }

    // &__favorite {
    //     border: none;
    //     background-color: transparent;
    //     cursor: pointer;
    // }

    // &__star {
    //     &.checked {
    //         color: #fb8e11cf !important;
    //     }
    // }
}