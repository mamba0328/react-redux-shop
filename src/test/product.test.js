import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from '@testing-library/user-event'
import Product from "../components/Product";

const product = {
    name: 'GIBSON LES PAUL TRIBUTE SATIN',
    price: '44 400',
    img: 'https://muztorg.ua/image/cache/catalog/99992/229497/229497_GIBSON%20LES%20PAUL%20TRIBUTE%20SATIN%20HONEYBURST_1-250x200.jpg',
    code: '228627',
    color: 'honey',
}

describe('Product', () => {
    it('Product matches snapshot', () => {
        render(<Product product={product} />);
        expect(<Product product={product} />).toMatchSnapshot();
    })

    it('Product should be visible', () => {
        render(<Product product={product} />);
        const productContent = screen.getByText('GIBSON LES PAUL TRIBUTE SATIN');
        expect(productContent).toBeInTheDocument();
    })

    it('Modal IN CART button toggles', () => {
        render(<Product product={product} inCart={true} />);
        const inCartButtom = screen.getByText('In cart');
        expect(inCartButtom).toBeInTheDocument();
    })

    it('Button triggers event', async () => {
        const user = userEvent.setup();
        const mockFunction = jest.fn(code => code);
        render(<Product product={product} addToCart={mockFunction} />);

        const addToCart = screen.getByText('Add to cart');
        await user.click(addToCart)
        expect(mockFunction).toHaveBeenCalled();
    })
})


