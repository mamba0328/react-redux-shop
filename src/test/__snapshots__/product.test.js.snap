// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Product matches snapshot 1`] = `
<Product
  inCart={false}
  isFavorite={false}
  product={
    Object {
      "code": "228627",
      "color": "honey",
      "img": "https://muztorg.ua/image/cache/catalog/99992/229497/229497_GIBSON%20LES%20PAUL%20TRIBUTE%20SATIN%20HONEYBURST_1-250x200.jpg",
      "name": "GIBSON LES PAUL TRIBUTE SATIN",
      "price": "44 400",
    }
  }
  xButton={false}
/>
`;
